adds cursor control ANSI escape codes

wassup05 opened the pull request:
The PR adds cursor control ANSI escape codes as a separate module `stdlib_ansi_cursor`

This was brought up scarcely in #229 but was not really discussed.

## Functions added are
1. `move_to(x,y)`
2. `move_to_column(x)`
3. `move_<direction>` where `<direction>` is any one of up, down, left, right
4. A couple of constants for clearning the terminal, clearning a line and it's different variants

## Motivation
There is already a module `stdlib_ansi` for ANSI escape codes of foreground, background colors and the different styles... If a person finds the use for such module it is very probable that he will also at some point want to control the cursor position too, to let's say, implement a custom loading animation or for elegant management of terminal output

### Side note
I chose to make it a different module because of the difference in approach between the two, the functions in this module talk in terms of strings whereas `stdlib_ansi` has a whole type surrounding it and also because you won't need to add the cursor_ to all the functions to make it more readable... But I don't really see a problem in merging the two too...

Edit: This is my first attempt at contributing here and I am not sure about who I am supposed to ask for feedback.. Could you guide me regarding this @perazz @jalvesz (I figured as you were the most active right now, hence pinging you would be appropriate)

wassup05 committed:
Add additional terminal escape sequences for cursor actions to expand the stdlib_ansi module

wassup05 committed:
Fix silly typo

jvdp1 commented on src/stdlib_ansi_cursor.f90:
```diff
@@ -0,0 +1,107 @@
+    character(len=*), parameter :: clear_entire_line = esc//"[2K"
+
+contains
+    !> moves the cursor to `(line, column)`
+    !> returns an empty string if any of them is negative
```
For each procedure, a `version` and a link to the specs should be provided.

jvdp1 commented on src/stdlib_ansi_cursor.f90:
```diff
@@ -0,0 +1,107 @@
+contains
+    !> moves the cursor to `(line, column)`
+    !> returns an empty string if any of them is negative
+    pure function move_to(line, col) result(str)
```
Could some of these functions be `elemental`?

wassup05 replied:
They could be made so, But I don't think it's worth it since I don't see anyone using it like that

jvdp1 commented on test/terminal/test_ansi_cursor.f90:
```diff
@@ -0,0 +1,90 @@
+        character(len=:), allocatable :: str
+
+        str = move_to(-10, 20)
+        call check(error, str, "")
+        if (allocated(error)) return
+        str = move_to(10, 20)
+        call check(error, iachar(str(1:1)), 27)
+        if (allocated(error)) return
+        call check(error, str(2:), "[10;20H")
```
Could you provide clear error messages, please?

jvdp1 reviewed (commented):
Thank you @wassup05 for this PR. It sounds like a good start. I left a few comments.

wassup05 committed:
fixed typo, added specs for stdlib_ansi_cursor, added helpful error messages to tests, small fixes

wassup05 commented on src/stdlib_ansi_cursor.f90:
```diff
@@ -0,0 +1,107 @@
+module stdlib_ansi_cursor
```
Thank you for reviewing!

It could be, but I thought separating it would be better since `stdlib_ansi` revolves around it's own derived type whereas this doesn't... I'm open to either way though

wassup05 commented on src/stdlib_ansi_cursor.f90:
```diff
@@ -0,0 +1,107 @@
+
+    character(len=*), parameter :: esc = achar(27)
+    !> moves the cursor to home => `(0,0)`
+    character(len=*), parameter :: home = esc//"[H"
```
Yes, It's more of a terminal thing than an OS thing.. But almost all of the modern terminals support these ansi codes

perazz commented on test/terminal/test_ansi_cursor.f90:
```diff
@@ -0,0 +1,120 @@
+        str = move_up(15)
+        call check(error, iachar(str(1:1)), 27)
+        if (allocated(error)) then
+            print *, "ERROR: move_up doesn't add ESC character at the beggining"
```
```suggestion
            print *, "ERROR: move_up doesn't add ESC character at the beginning"
```

perazz commented:
Nice work @wassup05, thank you. I'm no expert on ANSI terminal handling so I will add some of my thoughts here. 
- The `stdlib_ansi` module has a derived type for ANSI escape codes: 

https://github.com/fortran-lang/stdlib/blob/7511064e45476cc2deaa5af195fdb70cb73214d1/src/stdlib_ansi.f90#L23-L31

If the terminal handling codes provided here match the same pattern, I believe it would be better to uniform their handling with that of the module, and make them `type(ansi_code)` as well.
- Testing and examples are hard with console handling. However, I think it would be very useful to have examples that show some console interaction with these functions (for example, draw a box on the terminal, or something like that).

wassup05 committed:
typo fixes, small documentation changes and added an example

wassup05 commented:
Thanks for replying, @perazz !

When I started working on this, I thought of ways to integrate the two but couldn't really find any such way... As these are more action oriented whereas those of ansi_type were more of like changing the properties of the text being printed

Regarding examples... I added an example in the recent commit which basically draws a blue box on the screen using a subroutine `draw_box`, whose width, height and the position (line, column) were it is drawn can be changed.. I added the link to it in the spec as well

Please let me know what you think about it

14NGiestas committed:
Merge branch 'master' into ansi_codes
